modules/core/src/test/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImplSelfTest.java

```diff
@@ -110,6 +113,48 @@
         return cfg;
     }
 
+    /**
+     * @throws Exception If failed.
+     */
+    public void testCloseWithCancellation() throws Exception {
+        cnt = 0;
+
+        startGrids(3);
+
+        Ignite g2 = grid(2);
+
+        List<IgniteFuture> futures = new CopyOnWriteArrayList<>();
+
+        IgniteDataStreamer<Object, Object> dataLdr = g2.dataStreamer(DEFAULT_CACHE_NAME);
+
+        dataLdr.perNodeBufferSize(1);
+
+        multithreadedAsync(new Callable<Object>() {
+            @Override public Object call() throws Exception {
+                try {
+                    for (int i = 0; i < 100; i++)
+                        futures.add(dataLdr.addData(i, i));
+                }
+                catch (IllegalStateException ignored) {
+                    // No-op.
+                }
+                return null;
+            }
+        }, 3);
+
+        GridTestUtils.waitForCondition(new GridAbsPredicate() {
```

Comment:
Check result of `waitForCondition()`, because timout will return `false` value and not fail the test.